File: go/path_traversal/static_files.go
package pathtraversal

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
)

const staticDir = "/var/www/static"

// VULN 1: os.ReadDir with user-controlled directory - exposes arbitrary directory contents
func BrowseDirectory(w http.ResponseWriter, r *http.Request) {
	subdir := r.URL.Query().Get("dir")
	target := filepath.Join(staticDir, subdir)
	entries, err := os.ReadDir(target)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	var names []string
	for _, e := range entries {
		names = append(names, e.Name())
	}
	json.NewEncoder(w).Encode(names)
}

// VULN 2: os.ReadFile with fmt.Sprintf path - user controls name and type
func ServeAsset(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	assetType := r.URL.Query().Get("type")
	filePath := fmt.Sprintf("%s/%s/%s", staticDir, assetType, name)
	data, err := os.ReadFile(filePath)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	w.Write(data)
}

// VULN 3: os.WriteFile with user-controlled destination - write to arbitrary path
func CopyTemplate(w http.ResponseWriter, r *http.Request) {
	template := r.URL.Query().Get("template")
	dest := r.URL.Query().Get("dest")
	src := filepath.Join(staticDir, "templates", template)
	data, err := os.ReadFile(src)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	os.WriteFile(dest, data, 0644)
	w.Write([]byte("copied"))
}

File: python/path_traversal/static_files.py
import os
import shutil
from flask import Flask, request, jsonify

app = Flask(__name__)
STATIC_DIR = "/var/www/static"

# VULN 1: os.listdir with user-controlled directory - exposes arbitrary directory contents
@app.route("/browse")
def browse_directory():
    subdir = request.args.get("dir", "")
    target = os.path.join(STATIC_DIR, subdir)
    entries = os.listdir(target)
    return jsonify(entries)


# VULN 2: open() with f-string path - user controls file extension and name
@app.route("/asset")
def serve_asset():
    asset_name = request.args.get("name")
    asset_type = request.args.get("type")
    path = f"{STATIC_DIR}/{asset_type}/{asset_name}"
    with open(path, "rb") as f:
        return f.read()


# VULN 3: shutil.copyfile with user-controlled destination - write to arbitrary path
@app.route("/copy")
def copy_template():
    template = request.args.get("template")
    dest = request.args.get("dest")
    src_path = os.path.join(STATIC_DIR, "templates", template)
    shutil.copyfile(src_path, dest)
    return "copied"
